### ISSUE TITLE
Number of child processes

### ISSUE THREAD

**Issue by radutopala** [ochinchina/supervisord#38]
I have an issue with `numprocs`.

If I try to launch more than 200 instances (eg. 201) of the same command at once, it stalls and the `ctl status` does not return anything anymore. But it works well with <=200 instances.

The same trial with the Py supervisor works fine, I can launch for example 1000 instances of the same command.

**Comment by ochinchina** [comment]
make a fix for this issue and can you try it again?

**Comment by radutopala** [comment]
@ochinchina you already made it in master?

**Comment by ochinchina** [comment]
Tested with 2000 processes in my local Linux machine and already merged to the master.

I don't know if your OS is Linux. If it is not Linux and any issue found, please mail me.

**Comment by ochinchina** [comment]
@radutopala, Have you tested your case with more than 200 processes?

**Comment by radutopala** [comment]
Yes, still doesn't work as expected.. with what binary have you tested
yourself?

- Radu Topala

**Comment by ochinchina** [comment]
Yes, the binary I tested.

Can you show me your supervisor configuration? and I want to test it with your configuration

**Comment by radutopala** [comment]
```
[inet_http_server]
port=127.0.0.1:9001

[program:gnatsd]
command=/root/earthbot/bin/gnatsd -a localhost
process_name=%(program_name)s_%(process_num)02d
priority=1
numprocs=1
directory=/root/earthbot/
autostart=true
autorestart=true
stdout_logfile=/root/earthbot/logs/gnatsd.log
stdout_logfile_maxbytes=10MB
stderr_logfile=/root/earthbot/logs/gnatsd.log
stderr_logfile_maxbytes=10MB

[program:earthbot]
command=/root/earthbot/bin/earthbot -root=/root/earthbot
process_name=%(program_name)s_%(process_num)02d
priority=2
numprocs=1
directory=/root/earthbot/
autostart=true
autorestart=true
stdout_logfile=/root/earthbot/logs/earthbot.log
stdout_logfile_maxbytes=10MB
stderr_logfile=/root/earthbot/logs/earthbot.log
stderr_logfile_maxbytes=10MB

[program:worker-facebook]
command=/root/earthbot/bin/worker -root=/root/earthbot facebook
process_name=%(program_name)s_%(process_num)02d
priority=3
numprocs=200
directory=/root/earthbot/
autostart=true
autorestart=true
stdout_logfile=/root/earthbot/logs/worker-facebook.log
stdout_logfile_maxbytes=100MB
stderr_logfile=/root/earthbot/logs/worker-facebook.log
stderr_logfile_maxbytes=100MB
```

**Comment by radutopala** [comment]
@ochinchina when I was asking above about what binary have you used for testing, I meant what program have you run in the config with more than 200 processes?

**Comment by ochinchina** [comment]
I reproduced this issue in my test environment. 

This issue is caused by the configuration stdout_logfile and stderr_logfile. If the stdout_logfile or stderr_logfile is configured, too many file is opened and the file handler in the linux is used up.

So if set the open file handler to a big value, such like 102400 with linux command "ulimit -n 102400", 200 programs can be started.

**Comment by ochinchina** [comment]
could you please change configure your open file descriptor to a big value? and then re-test your case if it can be solved or not.

**Comment by radutopala** [comment]
I'll have some time to check soon.

- Radu Topala, Trisoft

**Comment by radutopala** [comment]
@ochinchina tried it out, have set in `/etc/security/limits.conf`:
```
* soft nofile 1024000
* hard nofile 1024000
* soft nproc 1024000
* hard nproc 1024000
```
I've managed to launch ~1200 processes like this. Do you have any idea how to extend this? Or is it just OS/hardware related now?..

**Comment by ochinchina** [comment]
Two improvements needed for this issue:

1) this limitation can also be set by the supervisor itself if user sets
the parameter "minfds" in the supervisord section (not implemented)
2) Need to refactor the XML RPC part to fast get the process status when
the supervisor starts applications.

2017-10-12 2:50 GMT+08:00 Radu Topala <notifications@github.com>:

**Comment by radutopala** [comment]
I think the problem is also hardware related. I've launched these ~1200 processes on a virtual machine with Ubuntu 16.04 and 4GB of RAM/4 vCPU allocated. All these processes occupied ~3GB of RAM while waiting for reqs. And the vCPUs were at max 100% while loading the processes, and then lowered to < 10%. 
Therefore, it is impossible to launch too many processes if you don't have enough RAM and CPU power, and this variates of course also by the amount of resources that a single process uses. Not sure if this can be calculated in a deterministic way, therefore may be more of a trial & error thing.

**Comment by ochinchina** [comment]
Agree, the number of processes can be launched by supervisor is related
with the available CPU&Memory resources. The supervisor can't use resources
exceed the resources provided by the bare metal/virtual machine.

2017-10-12 13:15 GMT+08:00 Radu Topala <notifications@github.com>:
